File: frontend/lib/junior.ts
import { api } from "./api";

export interface JuniorProfile {
  id: string;
  parent_id: string;
  first_name: string;
  age_range: "6-8" | "8-10" | "10-12";
  avatar: string;
  created_at: string;
  last_daily_claim: string | null;
  login_streak: number;
  last_login: string | null;
}

export interface JuniorMineral {
  mineral_id: string;
  name: string;
  family: string;
  rarity: "common" | "uncommon" | "rare" | "legendary";
  mohs_hardness: number;
  key_property: string;
  ontario_locality: string;
  flavour_text: string;
  fun_fact: string;
  age_easy_description: string;
  age_standard_description: string;
  detective_clues: string[];
  dig_site_associations: string[];
  province_associations: string[];
  card_emoji: string;
  card_colour: string;
}

export interface CollectionEntry {
  mineral: JuniorMineral;
  owned: boolean;
  unlocked_at: string | null;
  source: string | null;
}

export interface BadgeInfo {
  badge_id: string;
  name: string;
  category: "field" | "knowledge" | "engagement";
  description: string;
  icon: string;
  requirement: string;
  card_reward: string | null;
  earned: boolean;
  earned_at: string | null;
}

export interface DetectiveCaseInfo {
  case_id: string;
  title: string;
  region: string;
  difficulty: "easy" | "medium" | "hard";
  scene: string;
  clues: string[];
  options: string[];
  card_reward_rare: string;
  card_reward_uncommon: string;
  card_reward_common: string;
  unlocked: boolean;
  solved: boolean;
  attempts: number;
}

export interface DailyCardResult {
  mineral: JuniorMineral;
  is_new: boolean;
  streak: number;
  new_badges: string[];
}

export interface DetectiveSubmitResult {
  correct: boolean;
  explanation: string | null;
  correct_mineral: string | null;
  card_reward: string | null;
  attempts: number;
  new_badges: string[];
}

export interface ParentSummaryEntry {
  id: string;
  first_name: string;
  avatar: string;
  age_range: string;
  login_streak: number;
  cards_collected: number;
  badges_earned: number;
  cases_solved: number;
}

export const RARITY_COLORS: Record<string, string> = {
  common: "#94a3b8",
  uncommon: "#22c55e",
  rare: "#3b82f6",
  legendary: "#f59e0b",
};

export const RARITY_LABELS: Record<string, string> = {
  common: "Common",
  uncommon: "Uncommon",
  rare: "Rare",
  legendary: "Legendary ✨",
};

export const DIFF_COLORS: Record<string, string> = {
  easy: "bg-green-100 text-green-700",
  medium: "bg-amber-100 text-amber-700",
  hard: "bg-red-100 text-red-700",
};

export const juniorApi = {
  createProfile: (data: { first_name: string; age_range: string; avatar: string }) =>
    api.post<JuniorProfile>("/api/junior/profiles", data, { auth: true }),

  listProfiles: () =>
    api.get<JuniorProfile[]>("/api/junior/profiles", { auth: true }),

  updateProfile: (id: string, data: Partial<{ first_name: string; age_range: string; avatar: string }>) =>
    api.patch<JuniorProfile>(`/api/junior/profiles/${id}`, data, { auth: true }),

  deleteProfile: (id: string) =>
    api.del(`/api/junior/profiles/${id}`, { auth: true }),

  listMinerals: () =>
    api.get<JuniorMineral[]>("/api/junior/minerals"),

  getCollection: (juniorId: string) =>
    api.get<CollectionEntry[]>(`/api/junior/${juniorId}/collection`, { auth: true }),

  claimDailyCard: (juniorId: string) =>
    api.post<DailyCardResult>(`/api/junior/${juniorId}/daily-card`, {}, { auth: true }),

  getBadges: (juniorId: string) =>
    api.get<BadgeInfo[]>(`/api/junior/${juniorId}/badges`, { auth: true }),

  getDetectiveCases: (juniorId: string) =>
    api.get<DetectiveCaseInfo[]>(`/api/junior/${juniorId}/detective/cases`, { auth: true }),

  submitDetective: (juniorId: string, caseId: string, answerId: string) =>
    api.post<DetectiveSubmitResult>(
      `/api/junior/${juniorId}/detective/${caseId}/submit`,
      { answer_id: answerId },
      { auth: true }
    ),

  getParentSummary: () =>
    api.get<ParentSummaryEntry[]>("/api/junior/parent-summary", { auth: true }),
};

export function hasClaimedToday(profile: JuniorProfile): boolean {
  if (!profile.last_daily_claim) return false;
  const claimed = new Date(profile.last_daily_claim);
  const now = new Date();
  return (
    claimed.getFullYear() === now.getFullYear() &&
    claimed.getMonth() === now.getMonth() &&
    claimed.getDate() === now.getDate()
  );
}

File: backend/app/models/junior.py
from datetime import datetime
from enum import StrEnum

from beanie import Document, Indexed
from pydantic import Field
from pymongo import IndexModel


class AgeRange(StrEnum):
    YOUNG = "6-8"
    MIDDLE = "8-10"
    OLDER = "10-12"


class RarityTier(StrEnum):
    COMMON = "common"
    UNCOMMON = "uncommon"
    RARE = "rare"
    LEGENDARY = "legendary"


class BadgeCategory(StrEnum):
    FIELD = "field"
    KNOWLEDGE = "knowledge"
    ENGAGEMENT = "engagement"


class UnlockSource(StrEnum):
    DAILY = "daily"
    FIND_LOG = "find_log"
    BADGE = "badge"
    SITE_VISIT = "site_visit"
    STREAK = "streak"
    GAME = "game"
    BOOKING = "booking"


class DetectiveDifficulty(StrEnum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class JuniorProfile(Document):
    parent_id: Indexed(str)
    first_name: str
    age_range: AgeRange = AgeRange.MIDDLE
    avatar: str = "🪨"
    created_at: datetime = Field(default_factory=datetime.utcnow)
    last_daily_claim: datetime | None = None
    login_streak: int = 0
    last_login: datetime | None = None
    lessons_completed: list[str] = Field(default_factory=list)  # "track:lesson-slug"

    class Settings:
        name = "junior_profiles"
        indexes = [IndexModel([("parent_id", 1)])]


class JuniorMineral(Document):
    mineral_id: str  # unique slug e.g. "quartz"
    name: str
    family: str
    rarity: RarityTier
    mohs_hardness: float
    key_property: str
    ontario_locality: str
    flavour_text: str
    fun_fact: str
    age_easy_description: str
    age_standard_description: str
    detective_clues: list[str]
    dig_site_associations: list[str] = Field(default_factory=list)
    province_associations: list[str] = Field(default_factory=list)
    card_emoji: str = "💎"
    card_colour: str = "#e2e8f0"

    class Settings:
        name = "junior_minerals"
        indexes = [IndexModel([("mineral_id", 1)], unique=True)]


class JuniorCollection(Document):
    user_id: Indexed(str)
    junior_id: Indexed(str)
    mineral_id: str
    unlocked_at: datetime = Field(default_factory=datetime.utcnow)
    source: UnlockSource

    class Settings:
        name = "junior_collection"
        indexes = [
            IndexModel([("user_id", 1), ("junior_id", 1)]),
            IndexModel([("user_id", 1), ("junior_id", 1), ("mineral_id", 1)], unique=True),
        ]


class Badge(Document):
    badge_id: str
    name: str
    category: BadgeCategory
    description: str
    icon: str
    requirement: str
    card_reward: str | None = None  # mineral_id

    class Settings:
        name = "badges"
        indexes = [IndexModel([("badge_id", 1)], unique=True)]


class BadgeState(Document):
    user_id: Indexed(str)
    junior_id: Indexed(str)
    badge_id: str
    earned_at: datetime = Field(default_factory=datetime.utcnow)
    trigger_event: str

    class Settings:
        name = "badge_states"
        indexes = [
            IndexModel([("user_id", 1), ("junior_id", 1)]),
            IndexModel([("user_id", 1), ("junior_id", 1), ("badge_id", 1)], unique=True),
        ]


class DetectiveCase(Document):
    case_id: str
    title: str
    region: str
    difficulty: DetectiveDifficulty
    scene: str
    clues: list[str]
    options: list[str]  # 4 mineral names
    correct_id: str  # mineral_id
    explanation: str
    unlock_day: int  # cases unlock every 3 days; day 0 = always available
    card_reward_rare: str
    card_reward_uncommon: str
    card_reward_common: str

    class Settings:
        name = "detective_cases"
        indexes = [IndexModel([("case_id", 1)], unique=True)]


class DetectiveCaseHistory(Document):
    user_id: Indexed(str)
    junior_id: Indexed(str)
    case_id: str
    attempts: int = 0
    solved: bool = False
    completed_at: datetime | None = None

    class Settings:
        name = "detective_case_history"
        indexes = [
            IndexModel([("user_id", 1), ("junior_id", 1), ("case_id", 1)], unique=True),
        ]
